### ISSUE TITLE
Unable to leverage universal-cookie, react-cookie, and universal-cookie express with NodeNext module resolution

### ISSUE THREAD

**Issue by corbin-mosher** [bendotcodes/cookies#471]
We would like to use the `universal-cookie` dependencies in our packages that are fully ESM rather than CommonJS and leverage TypeScript as our build tool. The resulting packages are used by some Node applications as well as browser based applications generated with Vite.

However, the type definitions for the libraries do not seem to work with the `NodeNext` `moduleResolution` for TypeScript's `tsconfig`.

Here is a minimal reproduction of the issue: ([sandbox link](https://codesandbox.io/p/devbox/universal-cookie-esm-types-2x5ty3?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clucvgz1v00073b6fklychjcf%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clucvgz1v00023b6fps5gbu9s%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clucvgz1v00043b6fp4jrfxuk%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clucvgz1v00063b6fpbhupa4v%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clucvgz1v00023b6fps5gbu9s%2522%253A%257B%2522id%2522%253A%2522clucvgz1v00023b6fps5gbu9s%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clucvgz1v00063b6fpbhupa4v%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clucvgz1v00053b6ffzjcpkk9%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clucvgz1v00063b6fpbhupa4v%2522%252C%2522activeTabId%2522%253A%2522clucvgz1v00053b6ffzjcpkk9%2522%257D%252C%2522clucvgz1v00043b6fp4jrfxuk%2522%253A%257B%2522id%2522%253A%2522clucvgz1v00043b6fp4jrfxuk%2522%252C%2522activeTabId%2522%253A%2522clucvqhmd013v3b6fs0wgmd75%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clucvgz1v00033b6f0qgedzy9%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522install%2522%252C%2522id%2522%253A%2522clucvnfz000ln3b6flitk110i%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522build%2522%252C%2522id%2522%253A%2522clucvqhmd013v3b6fs0wgmd75%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D))
![Screenshot 2024-03-29 at 9 29 45 AM](https://github.com/bendotcodes/cookies/assets/91292058/271696e9-948d-452e-90d8-7670c9da9eea)

I believe the issue could maybe be similar to [this](https://github.com/microsoft/TypeScript/issues/50690#issuecomment-1241359555), but maybe not precisely. I am not super familiar with the challenges of exposing a package as a maintainer for ESM, CJS, and UMD consumption. So I am not 100% clear on how to alter the `universal-cookie` project to structure its types correctly for all three to work.

Thoughts on if this is something that can be patched?

**Comment by eXon** [comment]
Hi @corbin-mosher ,

The TypeScript types seem to be not working with your tsconfig configuration. I'm trying to dig deeper to find a solution to why, this is very unexpected.

**Comment by corbin-mosher** [comment]
@eXon, yeah, it is very confusing.

I even tried cloning the cookies project and running it through my normal build pattern and I have the same type issue.

The only way I have gotten it to work is to also export `Cookies` as a named export, not just the default. For some reason, with `NodeNext` tsconfig there is a struggle with this declared `Cookies` class, but only when its the default export 🙃 

There have been one or two more npm packages that have had similar issues for me with default exports and `NodeNext`, so I am likely going to just stick with `Bundler`+`ESNext` for now and continue using `unbuild` (rollup) to handle build commands.

I am super curious though to learn what causes this hangup when building and resolving things purely with `tsc`+`tsconfig`.

**Comment by eXon** [comment]
@corbin-mosher This is fixed in 7.1.2, let me know if it works for you

**Comment by corbin-mosher** [comment]
@eXon, thanks! Types are working as expected now with `NodeNext` for `moduleResolution` in `tsconfig`.

I really appreciate the quick help.

**Comment by corbin-mosher** [comment]
@eXon, now that I have had time to sit down with this further, it seems to me that the post build script plan might need some alterations.

It is impacting imports of external dependencies as well. For example, in `react-cookies` the type file for `CookieProvider` is altered from:
```
import * as React from 'react';
import Cookies from 'universal-cookie';
import { ReactCookieProps } from './types';
export default class CookiesProvider extends React.Component<ReactCookieProps, any> {
    cookies: Cookies;
    constructor(props: ReactCookieProps);
    render(): React.JSX.Element;
}
```

to
```
import * as React from 'react.d.mts';
import Cookies from 'universal-cookie.d.mts';
import { ReactCookieProps } from './types.d.mts';
export default class CookiesProvider extends React.Component<ReactCookieProps, any> {
    cookies: Cookies;
    constructor(props: ReactCookieProps);
    render(): React.JSX.Element;
}
```

Its possible that the change on line 3 is OK since it is referencing a file directly. But I think the changes on line 1 & 2 would cause issues. Let me know what you think.

**Comment by eXon** [comment]
@corbin-mosher You're right! I've fixed the regex, let me know if 7.1.4 works
